lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1613,30 +1624,72 @@ class HSAQueue final : public KalmarQueue
             printAsyncOps(std::cerr);
         }
 
+        std::lock_guard<std::recursive_mutex> lg(qmutex);
+
+#if !defined(NDEBUG)
+        bool foundFirstValidOp = false;
+#endif
+        int first_nullptr = -1;
+        int lastWaitOp = asyncOps.size();
+        if (drainingQueue_) {
+            lastWaitOp = (lastWaitOp * QUEUE_FLUSHING_FRAC);
+        }
 
-        {
-            std::lock_guard<std::recursive_mutex> lg(qmutex);
-            bool foundFirstValidOp = false;
-            for (int i = asyncOps.size()-1; i >= 0;  i--) {
-                if (asyncOps[i] != nullptr) {
-                    auto asyncOp = asyncOps[i];
-                    if (!foundFirstValidOp) {
-                        hsa_signal_t sig =  *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
-                        assert(sig.handle != 0);
-                        foundFirstValidOp = true;
-                    }
-                    // wait on valid futures only
-                    std::shared_future<void>* future = asyncOp->getFuture();
-                    if (future && future->valid()) {
-                        future->wait();
-                    }
+        for (int i = lastWaitOp-1; i >= 0;  i--) {
+            if (asyncOps[i] != nullptr) {
+                auto asyncOp = asyncOps[i];
+                // we only drain starting at first found marker
+                if (drainingQueue_ && asyncOp->getCommandKind() != hcCommandMarker) continue;
+#if !defined(NDEBUG)
```

Comment:
@jeffdaily could you clarify the rationale of this check? it seems `foundFirstValidOp` would NOT be set `true` in debug build?

Reply:
a deeper look suggest `foundFirstValidOp` is not really used anywhere else... should it be removed?

Reply:
`foundFirstValidOp` and its use was present before my changes.  I was trying to preserve the original code, but I noticed that `foundFirstValidOp` was only used to trigger the code path for the `assert(sig.handle != 0)`.  Perhaps it was premature optimization, but the `assert` would be a no-op if `NDEBUG` is defined, so I tried removing the entire use of `foundFirstValidOp` if `NDEBUG` was defined.

Reply:
I think the logic is to ensure `asyncOp->getFuture()` won't crash later on. From `git blame` it doesn't really seem to offer any specific help. Maybe we should remove the logic here.